File: bindings/node/e2e/lib/streaming-collector.ts
// Copyright 2024 mysql-event-stream Authors
// SPDX-License-Identifier: Apache-2.0

import { CdcStream } from "../../src/stream.js";
import type { ChangeEvent, ClientConfig, EventType } from "../../src/types.js";

interface WaitForEventsOptions {
  table?: string;
  type?: EventType;
  count?: number;
  timeout?: number;
}

/** Collects CDC events from a live MySQL binlog stream using CdcStream. */
export class StreamingCollector {
  private config: ClientConfig;
  private stream: CdcStream | null = null;
  private events: ChangeEvent[] = [];
  private stopped = false;
  private pollPromise: Promise<void> | null = null;
  // The stream's terminal error, held until a caller is in a position to see
  // it. Discarding it turns every connection, authentication or purged-GTID
  // failure into a bare count mismatch after the full wait, with the cause
  // gone.
  private streamError: Error | null = null;

  constructor(config: ClientConfig) {
    this.config = config;
  }

  /**
   * Start the binlog stream and begin collecting events.
   *
   * A configuration the stream refuses outright is raised from here, so it
   * surfaces at the call that made the mistake. A failure that needs the
   * network to happen -- a refused connection, a rejected login, a start
   * position the server has purged -- cannot be known before the first wait,
   * and is raised from {@link waitForEvents} instead.
   */
  async start(): Promise<void> {
    this.pollPromise = this.pollLoop();
    // Let the poll loop run up to its first suspension point, which is where a
    // refusal that needs no I/O has already landed.
    await Promise.resolve();
    this.throwIfStreamFailed();
  }

  /** Stop collecting and release resources. */
  async stop(): Promise<void> {
    this.stopped = true;
    if (this.stream) {
      await this.stream.close();
    }
    if (this.pollPromise) {
      await this.pollPromise;
    }
  }

  /**
   * Wait until matching events are collected.
   *
   * Polls the internal event buffer until enough matching events are found.
   *
   * @throws The stream's terminal error, as soon as one is seen -- a refused
   * connection or a rejected login ends the wait immediately rather than
   * running it out.
   * @throws If the deadline passes with fewer matching events than asked for.
   * Returning what it has would present a stream that never started as a
   * stream that produced nothing.
   */
  async waitForEvents(opts: WaitForEventsOptions = {}): Promise<ChangeEvent[]> {
    const { table, type, count = 1, timeout = 30_000 } = opts;
    const deadline = Date.now() + timeout;

    for (;;) {
      this.throwIfStreamFailed();
      const matched = this.matching(table, type);
      if (matched.length >= count) return matched;
      if (Date.now() >= deadline) {
        throw new Error(
          `Expected ${count} events (table=${table}, type=${type}) ` +
            `but got ${matched.length} within ${timeout}ms`,
        );
      }
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
  }

  /** Clear all collected events. */
  clearEvents(): void {
    this.events = [];
  }

  /** Collected events matching the optional table and type filters. */
  private matching(table?: string, type?: EventType): ChangeEvent[] {
    return this.events.filter((ev) => {
      if (table && ev.table !== table) return false;
      if (type && ev.type !== type) return false;
      return true;
    });
  }

  /** Re-raise the stream's terminal error, if it has one. */
  private throwIfStreamFailed(): void {
    if (this.streamError) throw this.streamError;
  }

  private async pollLoop(): Promise<void> {
    try {
      this.stream = new CdcStream(this.config);
      for await (const event of this.stream) {
        if (this.stopped) break;
        this.events.push(event);
      }
    } catch (err) {
      // A stop closes the stream out from under the iteration, so the error
      // that arrives then is the shutdown itself and carries nothing a test
      // wants. Anything raised while the collector still expects events is the
      // reason the events are not coming.
      if (!this.stopped) {
        this.streamError = err instanceof Error ? err : new Error(String(err));
      }
    }
  }
}

File: bindings/python/e2e/lib/streaming_collector.py
"""Streaming event collector for E2E tests.

Runs BinlogClient + CdcEngine in a background thread to collect
CDC events triggered by DML operations in the main thread.
"""

from __future__ import annotations

import threading
import time
from collections.abc import Callable

from mysql_event_stream import CdcEngine, ChangeEvent
from mysql_event_stream.client import BinlogClient


class StreamingCollector:
    """Collects CDC events from a BinlogClient stream.

    Usage::

        collector = StreamingCollector(lib_path, server_id=100)
        collector.start()
        try:
            mysql.insert("items", name="test", value=42)
            events = collector.wait_for_events(table="items", count=1)
            assert events[0].type == EventType.INSERT
        finally:
            collector.stop()
    """

    def __init__(
        self,
        lib_path: str,
        *,
        host: str = "127.0.0.1",
        port: int = 13308,
        user: str = "root",
        password: str = "test_root_password",
        server_id: int = 100,
        start_gtid: str = "",
    ) -> None:
        self._lib_path = lib_path
        self._host = host
        self._port = port
        self._user = user
        self._password = password
        self._server_id = server_id
        self._start_gtid = start_gtid
        self._events: list[ChangeEvent] = []
        self._lock = threading.Lock()
        self._stop = threading.Event()
        self._started = threading.Event()
        self._thread: threading.Thread | None = None
        self._error: Exception | None = None

    def start(self) -> None:
        """Start the streaming collector in a background thread."""
        self._stop.clear()
        self._started.clear()
        self._error = None
        self._events.clear()
        self._thread = threading.Thread(target=self._worker, daemon=True)
        self._thread.start()
        if not self._started.wait(timeout=30):
            raise RuntimeError("Streaming collector failed to start within 30s")
        if self._error:
            raise self._error

    def stop(self) -> None:
        """Stop the streaming collector."""
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=10)
            self._thread = None

    def wait_for_events(
        self,
        *,
        table: str | None = None,
        event_type: object = None,
        count: int = 1,
        timeout: float = 30.0,
        predicate: Callable[[ChangeEvent], bool] | None = None,
    ) -> list[ChangeEvent]:
        """Wait for matching events to arrive.

        Args:
            table: Filter by table name.
            event_type: Filter by EventType.
            count: Minimum number of matching events to collect.
            timeout: Maximum wait time in seconds.
            predicate: Custom filter function.

        Returns:
            List of matching ChangeEvent objects.

        Raises:
            TimeoutError: If not enough events arrive within timeout.
        """
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            with self._lock:
                matching = self._filter_events(table, event_type, predicate)
                if len(matching) >= count:
                    return matching[:count]
            time.sleep(0.05)

        with self._lock:
            matching = self._filter_events(table, event_type, predicate)
        raise TimeoutError(
            f"Expected {count} events (table={table}, type={event_type}) "
            f"but got {len(matching)} within {timeout}s"
        )

    @property
    def all_events(self) -> list[ChangeEvent]:
        """Get a snapshot of all collected events."""
        with self._lock:
            return list(self._events)

    def _filter_events(
        self,
        table: str | None,
        event_type: object,
        predicate: Callable[[ChangeEvent], bool] | None,
    ) -> list[ChangeEvent]:
        result = self._events
        if table:
            result = [e for e in result if e.table == table]
        if event_type is not None:
            result = [e for e in result if e.type == event_type]
        if predicate:
            result = [e for e in result if predicate(e)]
        return result

    def _worker(self) -> None:
        try:
            with BinlogClient(
                host=self._host,
                port=self._port,
                user=self._user,
                password=self._password,
                server_id=self._server_id,
                start_gtid=self._start_gtid,
                lib_path=self._lib_path,
            ) as client:
                client.connect()
                client.start()
                self._started.set()

                with CdcEngine(lib_path=self._lib_path) as engine:
                    engine.enable_metadata(
                        host=self._host,
                        port=self._port,
                        user=self._user,
                        password=self._password,
                    )
                    while not self._stop.is_set():
                        result = client.poll()
                        if result.is_heartbeat or not result.data:
                            continue
                        engine.feed(result.data)
                        while (ev := engine.next_event()) is not None:
                            with self._lock:
                                self._events.append(ev)
        except Exception as e:
            self._error = e
            self._started.set()
